src/collector.py

```diff
@@ -0,0 +1,451 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""OpenSearch Dashboards Collector."""
+
+import json
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from typing import Generator, Optional
+
+import requests
+from prometheus_client.core import GaugeMetricFamily, Metric
+from prometheus_client.registry import Collector
+from requests.auth import HTTPBasicAuth
+from requests.exceptions import HTTPError, RequestException, Timeout
+
+METRICS_PREFIX = "opensearch_dashboards_"
+API_STATUS_ENDPOINT = "/api/status"
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Config:
+    """Wrap CLI arguments and environment variables necessary to get metrics from the API."""
+
+    url: str
+    user: str
+    password: str
+
+
+class Heap(Enum):
+    """Possible heap types for the metrics"""
+
+    USED = "used_in_bytes"
+    TOTAL = "total_in_bytes"
+    SIZE = "size_limit"
+
+
+class Load(Enum):
+    """Possible heap types for the metrics"""
+
+    ONE_M = "1m"
+    FIVE_M = "5m"
+    FIFTEEN_M = "15m"
+
+
+class Memory(Enum):
+    """Possible memory types for the metrics"""
+
+    USED = "used_in_bytes"
+    TOTAL = "total_in_bytes"
+    FREE = "free_in_bytes"
+
+
+class Response(Enum):
+    """Possible response times types for the metrics"""
+
+    MAX = "max_in_millis"
+    AVG = "avg_in_millis"
+
+
+class RequestsCount(Enum):
+    """Possible requests count types for the metrics"""
+
+    DISCONNECTS = "disconnects"
+    TOTAL = "total"
+
+
+class DashBoardsCollector(Collector):
+    """OpenSearch Dashboards Collector"""
+
+    def __init__(self, config: Config) -> None:
+        super().__init__()
+        self.config = config
+
+    def collect(self) -> Generator[Metric, None, None]:
+        """Collect all the metrics from /api/status to be exported.
+
+        Yields:
+            Generator[Metric]: the internal status metrics of status
+        """
+        api_metrics = collect_api_status(self.config)
+        api_up = GaugeMetricFamily(
+            name=f"{METRICS_PREFIX}up",
+            documentation="Whether the data source is reachable (1 for up, 0 for down)",
+            value=1 if api_metrics else 0,
+        )
+        yield api_up
+
+        for metric in self.metrics(api_metrics):
+            description, value = metric
+            if value:
+                yield value
+            else:
+                logger.error("It was not possible to get the metric: %s", description)
+
+    def metrics(self, api_metrics: dict) -> list[tuple[str, Optional[Metric]]]:
+        """Get the OpenSearch dashboard prometheus metrics.
+
+        Args:
+            api_metrics (dict): Metrics from the OpenSearch Dashboards API
+
+        Returns:
+            list[tuple[str, Optional[Metric]]]: Prometheus Gauge metrics of the dashboards
+        """
+        if api_metrics:
```

Comment:
nit for clarity, given the size of the list: 
```suggestion
        if not api_metrics:
            return []
            
        return ....
```